custom/icds_reports/utils/__init__.py

```diff
@@ -892,7 +892,7 @@ def create_aww_performance_excel_file(excel_data, data_type, month, state, distr
     for column_index in range(len(columns)):
         widths[columns[column_index]] = max(
             widths[columns[column_index]],
-            max(len(str(row[column_index])) for row in excel_data[1:]) * 4 // 3 if len(excel_data) >= 2 else 0
+            max(len(unicode(row[column_index])) for row in excel_data[1:]) * 4 // 3 if len(excel_data) >= 2 else 0
```

Comment:
The operator ```unicode``` does not exist in python 3.  For compatibility, any instances of ```unicode``` should be replaced with ```six.text_type```.

The recommended pattern is to use ```.decode('utf-8')``` to turn bytes into text.
Calling ```six.text_type```/```str``` on bytes will give a different result in python 3:

Python 2:
```
>>> unicode(b'abc')
u'abc'
>>> import six
>>> six.text_type(b'abc')
u'abc'
>>> b'abc'.decode('utf-8')
u'abc'
```

Python 3:
```
>>> unicode(b'abc')
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
NameError: name 'unicode' is not defined
>>> str(b'abc')
"b'abc'"
>>> import six
>>> six.text_type(b'abc')
"b'abc'"
>>> b'abc'.decode('utf-8')
'abc'
```

Note that in Python 3, all string literals without a prefix default to type text.

We should also never guess whether data is unicode or not.  If necessary, check whether your variable is type ```bytes``` or ```six.text_type```.

Reply:
Thanks Nick, replace unicode with six.text_type.

Reply:
https://github.com/dimagi/commcare-hq/pull/22820/commits/eb29bd2803de3183a6a682db22630d850da2fded will also fail for non-ascii byte sequences because the default encoding is ascii.  The encoding should always be specified, since bytes don't mean anything on their own without an encoding.

```
>>> nonascii = b'ç'
>>> nonascii
'\xc3\xa7'
>>> import six
>>> six.text_type(nonascii)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
UnicodeDecodeError: 'ascii' codec can't decode byte 0xc3 in position 0: ordinal not in range(128)
>>> if not isinstance(nonascii, six.text_type):
...    print repr(nonascii.decode('utf-8'))
... 
u'\xe7'
```

Whenever in doubt, I recommend referencing https://nedbatchelder.com/text/unipain.html.  It's worth re-watching/re-reading until you can recite it in your sleep.